src/accountsdb/cache.zig

```diff
@@ -0,0 +1,209 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Pubkey = sig.core.pubkey.Pubkey;
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+const ReferenceCounter = sig.sync.reference_counter.ReferenceCounter;
+
+/// Stores read-only in-memory copies of commonly used *rooted* accounts
+pub const AccountsCache = struct {
+    lru: LRU,
+    allocator: std.mem.Allocator,
+
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        ref_count: ReferenceCounter,
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) error{OutOfMemory}!CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = .{},
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            if (self.ref_count.release()) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            if (self.ref_count.acquire()) {
+                return self;
+            } else {
+                @panic("Attempted to acquire reference to destroyed CachedAccount");
```

Comment:
This panic doesn't make sense to have, `ref_count` will also be destroyed with the last release.

Reply:
I agree that the only case where this could theoretically be false is when the pointer itself is already invalidated and unsafe to use. If the function is called at that point, it means you already have a bug somewhere else, from mismanaging the pointer, and the behavior of calling this function is undefined.

So you could just have `_ = self.ref_count.acquire()` here. Essentially you use it as a basic reference counter, and don't bother to check for cases that are actually impossible to accurately detect.

That being said, maybe there is an advantage to having a panic or unreachable here. Nothing is guaranteed, since any code that could trigger this would have undefined behavior, but at least you get a probabilistic tilt towards failing fast instead of failing silently.

Reply:
I went ahead and replaced it with an unreachable, I think that shows intent a bit better